util/downloader/downloader.go

```diff
@@ -182,83 +183,68 @@ func (d *Downloader) downloadFile(ctx context.Context, out *os.File, stats *Stat
 		_ = resp.Body.Close()
 	}()
 
-	buffer := make([]byte, 32*1024)
-	hasher := sha256.New()
-
-	if err := d.updateHasherWithExistingData(stats.Downloaded, hasher); err != nil {
-		return err
-	}
-
-	return d.writeToFile(ctx, resp, out, buffer, hasher, stats)
-}
-
-func (d *Downloader) createRequest(ctx context.Context, downloaded int64) (*http.Request, error) {
-	req, err := http.NewRequestWithContext(ctx, http.MethodGet, d.url, http.NoBody)
-	if err != nil {
-		return nil, ErrNewRequest
-	}
-	if downloaded > 0 {
-		req.Header.Set("Range", fmt.Sprintf("bytes=%d-", downloaded))
+	if resp.StatusCode != http.StatusPartialContent && resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("got http response %s from %s: %w", resp.Status, d.url, err)
 	}
 
-	return req, nil
-}
+	buf := make([]byte, 32*1024) // 32KB buffer for reading the response body
+	offset := c.start
+	for {
+		n, err := resp.Body.Read(buf)
+		if n > 0 {
+			d.mu.Lock()
+			for written := 0; written < n; {
+				w, err := out.WriteAt(buf[written:n], offset+int64(written))
+				if err != nil {
+					d.mu.Unlock()
 
-func (d *Downloader) updateHasherWithExistingData(downloaded int64, hasher io.Writer) error {
-	if downloaded > 0 {
-		existingFile, err := os.Open(d.filePath)
-		if err != nil {
-			return ErrOpenFileExists
+					return ErrFileWriting
```

Comment:
Better we return the `err` as it may have some useful information. You can forma error message.